README.md
# AIY Voice Only

[Google AIY Voice Kit](https://aiyprojects.withgoogle.com/voice) is a cool
project.  But it locks you into its custom hardware. I have separated its
software to work on Raspberry Pi (3B and 3B+) independently, just using a normal
speaker and microphone.

The following instructions aim at:

**Raspberry Pi (3B, 3B+)**  
**Raspbian Stretch**  
**Python 3**

Additionally, you need:

- a **Speaker** to plug into Raspberry Pi's headphone jack
- a **USB Microphone**

**Plug them in.** Let's go.

## Find your speaker and mic

Locate your speaker in the list of playback hardware devices. Normally, it is at
**card 0, device 0**, as indicated by the sample output below.

```
$ aplay -l

**** List of PLAYBACK Hardware Devices ****
card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

Locate your USB microphone in the list of capture hardware devices. Normally, it
is at **card 1, device 0**, as indicated by the sample output below.

```
$ arecord -l

**** List of CAPTURE Hardware Devices ****
card 1: Device [USB PnP Audio Device], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

*Your hardware's number might be different from mine. Adapt accordingly.*

## Make them the defaults

Create a new file named `.asoundrc` in the home directory (`/home/pi`). Put in
the following contents. Adjust the `card,device` number if needed.

```
pcm.!default {
  type asym
  capture.pcm "mic"
  playback.pcm "speaker"
}
pcm.mic {
  type plug
  slave {
    pcm "hw:1,0"   # card number, device number
  }
}
pcm.speaker {
  type plug
  slave {
    pcm "hw:0,0"   # card number, device number
  }
}
```

## Make sure sound output to headphone jack

Sound may be output via HDMI or headphone jack. We want to use the headphone
jack.

Enter `sudo raspi-config`. Select **Advanced Options**, then **Audio**. Choose
**Force 3.5mm (headphone) jack**.

## Turn up the volume

A lot of times when sound applications seem to fail, it is because we forget to
turn up the volume.

Volume adjustment can be done with `alsamixer`. This program makes use of some
function keys (`F1`, `F2`, etc). For function keys to function properly on
PuTTY, we need to change some settings (click on the top-left corner of the
PuTTY window, then select **Change Settings ...**):

1. Go to **Terminal** / **Keyboard**
2. Look for section **The Function keys and keypad**
3. Select **Xterm R6**
4. Press button **Apply**

Now, we are ready to turn up the volume, for both the speaker and the mic:

```
$ alsamixer
```
`F6` to select between sound cards  
`F3` to select playback volume (for speaker)  
`F4` to select capture volume (for mic)  
`⬆` `⬇` arrow keys to adjust  
`Esc` to exit

*If you unplug the USB microphone at any moment, all volume settings
(including that of the speaker) may be reset. Make sure to check the volume
again.*

Hardware all set, let's test them.

## Test the speaker

```
$ speaker-test -t wav
```

Press `Ctrl-C` when done.

## Record a WAV file

```
$ arecord --format=S16_LE --duration=5 --rate=16000 --file-type=wav out.wav
```

## Play a WAV file

```
$ aplay out.wav
```

## Register for Google Assistant or Google Cloud Speech

Although we are not using Google's hardware, there is no escaping from its
software. We still rely on Google Assistant or Google Cloud Speech API to
perform voice recognition. To use these cloud services, you have to go through a
series of registration steps:

- [Configure Google Assistant API](https://developers.google.com/assistant/sdk/guides/library/python/embed/config-dev-project-and-account)
- [Configure Google Cloud Speech API](https://aiyprojects.withgoogle.com/voice#makers-guide-3-1--change-to-the-cloud-speech-api)

Which one to use depends on what you need. **Google Assistant** can recognize
speech *and* talk back intelligently, but supports fewer languages. **Google
Cloud Speech** only recognizes speech (no talk-back), but supports far more
languages.

Here is a summary of the steps for using **Google Assistant**, as of 2019-11-27.
Always pay attention to the **Project ID** on top of the page to make sure you
are in the right project.

1. Create a Project

2. Enable Google Assistant API

3. Configure OAuth consent screen (must fill in **Support email**)

4. Enable activity controls

5. Register device model, Download credentials file (check `project_id`)

6. Install system dependencies:
    ```
    $ sudo apt-get install portaudio19-dev libffi-dev libssl-dev
    ```

7. Install Python packages:
    ```
    $ sudo pip3 install --upgrade pip setuptools wheel
    $ sudo pip3 install google-assistant-library==1.0.1 \
                        google-assistant-grpc==0.2.0 \
                        google-assistant-sdk[samples]==0.5.1 \
                        google-auth-oauthlib[tool] \
                        google-cloud-speech
    ```
   On Raspbian Buster, you likely have to downgrade to `google-assistant-library==1.0.0`
   for it to work.

8. Use `google-oauthlib-tool` to authenticate:
   ```
   $ google-oauthlib-tool --scope https://www.googleapis.com/auth/assistant-sdk-prototype \
      --save --headless --client-secrets /path/to/client_secret_client-id.json
   ```

9. Use [`googlesamples-assistant-devicetool`](https://developers.google.com/assistant/sdk/reference/device-registration/device-tool)
   to register your Raspberry Pi. A few useful commands may be:
   ```
   $ googlesamples-assistant-devicetool --project-id <Project ID> register-device \
   --model <Model ID> \
   --device <Make up a new Device ID> \
   --client-type LIBRARY

   $ googlesamples-assistant-devicetool --project-id <Project ID> list --model

   $ googlesamples-assistant-devicetool --project-id <Project ID> list --device
   ```

## How to use this library

I used to have it uploaded to PYPI for easy installation. But Google Assistant
is changing too rapidly. I find it more informing to download and try to
integrate it manually:

1. Download the `aiy` directory

2. Set environment variable `PYTHONPATH` so Python can find the `aiy` package

3. You may have to install the Pico text-to-speech engine, `libttspico-utils`,
   to allow it to generate speech dynamically (If `apt-get install libttspico-utils`
   doesn't work, manually download and install deb packages `libttspico-data`,
   `libttspico0`, and `libttspico-utils` separately)

The best way to experience the software is to try it.
**[Let's go to the examples.](https://github.com/nickoala/aiy-voice-only/tree/aiyprojects/examples/voice)**

## Changes to original library

Here is an outline of the changes I have made to the original [AIY Voice
Kit](https://github.com/google/aiyprojects-raspbian) source code:

1. No Vision stuff: The AIY project actually includes the [Vision
Kit](https://aiyprojects.withgoogle.com/vision) and associated software, which
are of no concern to this project. I have removed those.

2. No Voice Hat stuff: This project does not rely on the Voice Hat. The
`aiy.board` module has been removed.

3. The class `Led` and `Button` have been moved to the `aiy.util` module.


CHANGES.md
# Change log

This page describes the changes in each release.

To update your kit, see the [system updates guide][system-updates].
All system images can be downloaded from the [GitHub releases page][github-releases].

## AIY Kits Release 2021-04-02

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

Based on the [Raspberry Pi OS with desktop][raspberry-pi-os] from March 4th 2021.

**Fixes**

* Sound driver compilation errors on the latest Raspberry Pi OS
* Upgrade Colab TF version to 1.13.1
* Driver source code added to this repository

## AIY Kits Release 2020-11-20

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

Based on the [Raspberry Pi OS with desktop][raspberry-pi-os] from August 20th 2020.

**Fixes**

* Driver compilation errors on the latest Raspberry Pi OS
* HACKING.md instructions for Vision Bonnet and Voice HAT

## AIY Kits Release 2019-11-13

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

Based on the [Raspbian Buster with desktop][raspbian].

**Fixes**

* Fix driver compilation errors on the latest Raspbian
* Fix HACKING.md instructions
* Fix gpiozero integration

**Improvements**

* Debian packages are now hosted at https://packages.cloud.google.com/
* Original Raspbian image is modified in the minimal way (no unnecessary packages)

## AIY Kits Release 2018-11-16

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

**Fixes**

* Fix tts engine click sound
* Fix `assistant_grpc_demo.service`: add DISPLAY environment variable and proper
systemd dependencies
* Fix various linter findings: python3 compatibility, wrong variable names, etc.

**Improvements**

* New `board.py` to access button and LED on all boards
* New audio API in `voice/audio.py`
* Direct python support for iNaturalist models
* Load anchors/labels directly from text files
* Add `get_inference_state()` and `resest()` to Vision Bonnet protocol
* Add Voice HAT schematic in `docs` folder
* Add sparse representation for output tensors to increase data transfer rate
* SVG-image overlay for video streaming (experimental)

## AIY Kits Release 2018-08-03

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

**Fixes**

* Fix PulseAudio infinite loop with Voice Bonnet
* Fix PulseAudio volume control
* Fix gpiozero LED on/off bug
* Fix local USB networking on macOS, no driver required
* Fix check_audio.py

**Improvements**

* Add Makefile for common shortcuts
* Add vision unit tests for all models and examples
* Add video streaming support (experimental)
* Add Google Cloud IoT support (experimental)
* Add more documentation (pinouts, drivers, troubleshooting, etc.)
* Add new code examples and update existing ones
* Add CHANGES.md to track release changes
* Remove unnecessary files (e.g. ALSA configs)
* Update vision driver to support mmap syscall
* Update sound driver to support latest Raspbian image
* Update HACKING.md

## AIY Kits Release 2018-04-13

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

## AIY Kits Release 2018-02-21

Compatible with Voice HAT, Voice Bonnet, and Vision Bonnet.

## AIY Kits Release 2017-12-18

Compatible with Voice HAT and Vision Bonnet.

## VoiceKit Classic Image 2017-09-11

Compatible with Voice HAT.

[github-releases]: https://github.com/google/aiyprojects-raspbian/releases
[system-updates]: https://github.com/google/aiyprojects-raspbian/blob/aiyprojects/HACKING.md
[raspbian]: https://www.raspberrypi.org/downloads/raspbian/
[raspberry-pi-os]: https://www.raspberrypi.org/software/operating-systems/


src/main.py
#!/usr/bin/env python3
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Run a recognizer using the Google Assistant Library.

The Google Assistant Library has direct access to the audio API, so this Python
code doesn't need to record audio. Hot word detection "OK, Google" is supported.

The Google Assistant Library can be installed with:
    env/bin/pip install google-assistant-library==0.0.2

It is available for Raspberry Pi 2/3 only; Pi Zero is not supported.
"""

import logging
import subprocess
import sys
import time
import json

import os.path
import pathlib2 as pathlib
import configargparse

import google.oauth2.credentials

# from google.assistant.library import Assistant
# from google.assistant.library.file_helpers import existing_file
# from google.assistant.library.device_helpers import register_device
from google.assistant.library.event import EventType

from aiy.assistant import auth_helpers
from aiy.assistant.library import Assistant
from aiy.board import Board, Led
from aiy.voice import tts

from modules.kodi import KodiRemote
from modules.music import Music, PodCatcher
from modules.readrssfeed import ReadRssFeed
from modules.mqtt import MQTTSwitch
from modules.powercommand import PowerCommand

_configPath = os.path.expanduser('~/.config/voice-assistant.ini')
_settingsPath = os.path.expanduser('~/.config/settings.ini')
_remotePath = "https://argos.int.mpember.net.au/rpc/get/remotes"

_kodiRemote = KodiRemote(_settingsPath)
_music = Music(_settingsPath)
_podCatcher = PodCatcher(_settingsPath)
_readRssFeed = ReadRssFeed(_settingsPath)
_MQTTSwitch = MQTTSwitch(_settingsPath, _remotePath)

try:
    FileNotFoundError
except NameError:
    FileNotFoundError = IOError


WARNING_NOT_REGISTERED = """
    This device is not registered. This means you will not be able to use
    Device Actions or see your device in Assistant Settings. In order to
    register this device follow instructions at:

    https://developers.google.com/assistant/sdk/guides/library/python/embed/register-device
"""

def _createPID(pid_file='voice-recognizer.pid'):

    pid_dir = '/run/user/%d' % os.getuid()

    if not os.path.isdir(pid_dir):
        pid_dir = '/tmp'

    logging.info('PID stored in ' + pid_dir)

    file_name = os.path.join(pid_dir, pid_file)
    with open(file_name, 'w') as pid_file:
        pid_file.write("%d" % os.getpid())

def _volumeCommand(change):

    """Changes the volume and says the new level."""

    res = subprocess.check_output(r'amixer get Master | grep "Front Left:" | sed "s/.*\[\([0-9]\+\)%\].*/\1/"', shell=True).strip()
    try:
        logging.info('volume: %s', res)
        if change == 0 or change > 10:
            vol = change
        else:
            vol = int(res) + change

        vol = max(0, min(100, vol))
        if vol == 0:
            tts.say('Volume at %d %%.' % vol)

        subprocess.call('amixer -q set Master %d%%' % vol, shell=True)
        tts.say('Volume at %d %%.' % vol)

    except (ValueError, subprocess.CalledProcessError):
        logging.exception('Error using amixer to adjust volume.')

def process_event(assistant, led, event):

    global _cancelAction

    logging.info(event)
    if event.type == EventType.ON_START_FINISHED:
        led.state = Led.OFF  # Ready.
        print('Say "OK, Google" then speak, or press Ctrl+C to quit...')
        

    elif event.type == EventType.ON_CONVERSATION_TURN_STARTED:
        led.state = Led.ON

    elif event.type == EventType.ON_ALERT_STARTED and event.args:
        logging.warning('An alert just started, type = ' + str(event.args['alert_type']))
        assistant.stop_conversation()

    elif event.type == EventType.ON_ALERT_FINISHED and event.args:
        logging.warning('An alert just finished, type = ' + str(event.args['alert_type']))

    elif event.type == EventType.ON_RECOGNIZING_SPEECH_FINISHED and event.args:

        _cancelAction = False
        text = event.args['text'].lower()

        logging.info('You said: ' + text)

        if text == '':
            assistant.stop_conversation()

        elif _music.getConfirmPlayback() == True:
            assistant.stop_conversation()
            if text == 'yes':
                _music.command('podcast', 'CONFIRM')
            else:
                _music.setConfirmPlayback(False)
                _music.setPodcastURL(None)

#        elif text.startswith('music '):
#            assistant.stop_conversation()
#            _music.command('music', text[6:])

        elif text.startswith('podcast '):
            assistant.stop_conversation()
            _music.command('podcast', text[8:], _podCatcher)
            if _music.getConfirmPlayback() == True:
                assistant.start_conversation()

        elif text.startswith('play ') and text.endswith(' podcast'):
            assistant.stop_conversation()
            _music.command('podcast', text[5:][:-8], _podCatcher)
            if _music.getConfirmPlayback() == True:
                assistant.start_conversation()

        elif text.startswith('play ') and text.endswith(' podcasts'):
            assistant.stop_conversation()
            _music.command('podcast', text[5:][:-9], _podCatcher)
            if _music.getConfirmPlayback() == True:
                assistant.start_conversation()

        elif text.startswith('radio '):
            assistant.stop_conversation()
            _music.command('radio', text[6:])

        elif text.startswith('headlines '):
            assistant.stop_conversation()
            _readRssFeed.run(text[10:])

        elif text.startswith('turn on ') or text.startswith('turn off ') or text.startswith('turn down ') or text.startswith('turn up '):
            assistant.stop_conversation()
            _MQTTSwitch.run(text[5:])

        elif text.startswith('switch to channel '):
            assistant.stop_conversation()
            _kodiRemote.run('tv ' + text[18:])

        elif text.startswith('switch '):
            assistant.stop_conversation()
            _MQTTSwitch.run(text[7:])

        elif text.startswith('media center '):
            assistant.stop_conversation()
            _kodiRemote.run(text[13:])

        elif text.startswith('kodi ') or text.startswith('cody '):
            assistant.stop_conversation()
            _kodiRemote.run(text[5:])

        elif text.startswith('play the next episode of '):
            assistant.stop_conversation()
            _kodiRemote.run('play unwatched ' + text[25:])

        elif text.startswith('play next episode of '):
            assistant.stop_conversation()
            _kodiRemote.run('play unwatched ' + text[21:])

        elif text.startswith('play the most recent episode of '):
            assistant.stop_conversation()
            _kodiRemote.run('play unwatched ' + text[32:])

        elif text.startswith('play most recent episode of '):
            assistant.stop_conversation()
            _kodiRemote.run('play unwatched ' + text[28:])

        elif text.startswith('play unwatched ') or text.startswith('play tv series '):
            assistant.stop_conversation()
            _kodiRemote.run(text)

        elif text.startswith('play the lastest recording of '):
            assistant.stop_conversation()
            _kodiRemote.run('play recording ' + text[31:])

        elif text.startswith('play the recording of '):
            assistant.stop_conversation()
            _kodiRemote.run('play recording ' + text[23:])

        elif text.startswith('play recording of '):
            assistant.stop_conversation()
            _kodiRemote.run('play recording ' + text[18:])

        elif text.startswith('play recording '):
            assistant.stop_conversation()
            _kodiRemote.run(text)

        elif text.startswith('tv '):
            assistant.stop_conversation()
            _kodiRemote.run(text)

        elif text in ['power off','shutdown','shut down','self destruct']:
            assistant.stop_conversation()
            PowerCommand('shutdown')

        elif text in ['restart', 'reboot']:
            assistant.stop_conversation()
            PowerCommand('reboot')

        elif text == 'volume up':
            assistant.stop_conversation()
            _volumeCommand(10)

        elif text == 'volume down':
            assistant.stop_conversation()
            _volumeCommand(-10)

        elif text == 'volume maximum':
            assistant.stop_conversation()
            _volumeCommand(100)

        elif text == 'volume mute':
            assistant.stop_conversation()
            _volumeCommand(0)

        elif text == 'volume reset':
            assistant.stop_conversation()
            _volumeCommand(80)

        elif text == 'volume medium':
            assistant.stop_conversation()
            _volumeCommand(50)

        elif text == 'volume low':
            assistant.stop_conversation()
            _volumeCommand(30)

        elif text in ['brightness low', 'set brightness low']:
            assistant.stop_conversation()
            led.brightness(0.4)

        elif text in ['brightness medium', 'set brightness medium']:
            assistant.stop_conversation()
            led.brightness(0.7)

        elif text in ['brightness high', 'set brightness high', 'brightness full', 'set brightness full']:
            assistant.stop_conversation()
            led.brightness(1)

    elif event.type == EventType.ON_RESPONDING_FINISHED:
        assistant.stop_conversation()
        logging.info('EventType.ON_RESPONDING_FINISHED')

    elif event.type == EventType.ON_MEDIA_TRACK_LOAD:
        assistant.stop_conversation()
        logging.info('EventType.ON_MEDIA_TRACK_LOAD')
        logging.info(event.args)

    elif event.type == EventType.ON_MEDIA_TRACK_PLAY:
        assistant.stop_conversation()
        logging.info('EventType.ON_MEDIA_TRACK_PLAY')
        logging.info(event.args)

    elif event.type == EventType.ON_END_OF_UTTERANCE:
        led.state = Led.PULSE_QUICK

    elif (event.type == EventType.ON_CONVERSATION_TURN_FINISHED
          or event.type == EventType.ON_CONVERSATION_TURN_TIMEOUT
          or event.type == EventType.ON_NO_RESPONSE):
        led.state = Led.OFF

    elif event.type == EventType.ON_ASSISTANT_ERROR and event.args and event.args['is_fatal']:
        logging.info('EventType.ON_ASSISTANT_ERROR')
        sys.exit(1)

    elif event.type == EventType.ON_ASSISTANT_ERROR:
        logging.info('EventType.ON_ASSISTANT_ERROR')

    elif event.args:
        logging.info(event.type)
        logging.info(event.args)

    else:
        logging.info(event.type)

def main():

    parser = configargparse.ArgParser(
        default_config_files=[_configPath],
        description="Act on voice commands using Google's speech recognition")
    parser.add_argument('-L', '--language', default='en-GB',
                        help='Language code to use for speech (default: en-GB)')
    parser.add_argument('-p', '--pid-file', default='voice-recognizer.pid',
                        help='File containing our process id for monitoring')
    parser.add_argument('--trigger-sound', default=None,
                        help='Sound when trigger is activated (WAV format)')
    parser.add_argument('--brightness-max', type=int, default=1,
                        help='Maximum LED brightness')
    parser.add_argument('--brightness-min', type=int, default=1,
                        help='Minimum LED brightness')
    parser.add_argument('-d', '--daemon', action='store_true',
                        help='Daemon Mode')

    args = parser.parse_args()

    _createPID(args.pid_file)

    if args.daemon is True or sys.stdout.isatty() is not True:
        _podCatcher.start()
        logging.info("Starting in daemon mode")
    else:
        logging.info("Starting in non-daemon mode")

    credentials = auth_helpers.get_assistant_credentials()
    with Board() as board, Assistant(credentials) as assistant:
        logging.info('Setting brightness to %d %%' % args.brightness_max)
        logging.info(board.led)
        logging.info('Setting brightness to %d %%' % args.brightness_max)
        # led.brightness(0.2)

        for event in assistant.start():
            process_event(assistant, board.led, event)

if __name__ == '__main__':
    try:
        if sys.stdout.isatty():
            logging.basicConfig(
                level=logging.INFO,
                format="%(levelname)s:%(name)s:%(message)s"
            )
        else:
            logging.basicConfig(
                level=logging.WARNING,
                format="%(levelname)s:%(name)s:%(message)s"
            )

        main()
    except:
        pass

